File: Stone Game II.py
# https://leetcode.com/problems/stone-game-ii/?envType=daily-question&envId=2024-08-20
# Stone Game II

class Solution:
    def stoneGameII(self, piles: List[int]) -> int:

        n = len(piles)
    
        suffix_sums = [0] * (n + 1)

        for i in range(n - 1, -1, -1):
            suffix_sums[i] = suffix_sums[i + 1] + piles[i]
        
        dp = [[0] * (n + 1) for _ in range(n)]
        
        def dfs(i, m):
            if i >= n:
                return 0
            if dp[i][m] != 0:
                return dp[i][m]
            
            max_stones = 0

            for x in range(1, 2 * m + 1):
                if i + x > n:
                    break
                max_stones = max(max_stones, suffix_sums[i] - dfs(i + x, max(m, x)))
            
            dp[i][m] = max_stones

            return max_stones
        
        return dfs(0, 1)

File: Stone Game II.js
// https://leetcode.com/problems/stone-game-ii/?envType=daily-question&envId=2026-08-09
// Stone Game II

/**
 * @param {number[]} piles
 * @return {number}
 */
var stoneGameII = function(piles) {
    
    const n = piles.length;

    const memo = Array.from(
        { length: n + 1 },
        () => Array.from(
            { length: n + 1 },
            () => [-1, -1]
        )
    );

    function solve(i, m, t) {
        if (i >= n) return 0;

        if (memo[i][m][t] !== -1) {
            return memo[i][m][t];
        }

        let stones = 0;
        let res = t === 0 ? -1 : Infinity;

        for (let x = 1; x <= 2 * m; x++) {
            if (i + x - 1 >= n) break;

            stones += piles[i + x - 1];

            if (t === 0) {
                res = Math.max(
                    res,
                    stones + solve(i + x, Math.max(x, m), 1)
                );
            } else {
                res = Math.min(
                    res,
                    solve(i + x, Math.max(x, m), 0)
                );
            }
        }

        memo[i][m][t] = res;
        return res;
    }

    return solve(0, 1, 0);

};